File: Project5_makhan25/LinkedList.c
/**
 * This file implements a polymorphic linked list in C that is capable of storing any data type.
 * The list's operations are designed to be generic, allowing for various data-specific manipulations
 * via function pointers. This implementation is especially focused on ensuring proper memory management 
 * to prevent memory leaks and to provide basic list functionalities such as add, remove, and search.
 *
 * Functions Defined:
 * - ll_create: Initializes a new empty linked list.
 * - ll_push: Adds a new node with provided data to the front of the list.
 * - ll_pop: Removes and returns the data from the node at the front of the list.
 * - ll_append: Adds a new node with provided data to the end of the list.
 * - ll_remove: Removes a node with matching data based on a provided comparison function.
 * - ll_delete: Removes a node at a specified position.
 * - ll_find: Searches for a node with the given data using a comparison function.
 * - ll_size: Returns the number of nodes in the list.
 * - ll_clear: Clears the list, freeing all associated memory.
 * - ll_map: Applies a function to the data in each node.
 *
 * Author: Mahdeen Ahmed Khan Sameer
 * Date: April 10, 2024
 */

#include <stdlib.h>
#include "linkedlist.h"

/* Creates and initializes a new linked list */
LinkedList *ll_create() {
    LinkedList *list = (LinkedList *)malloc(sizeof(LinkedList));
    if (list != NULL) {
        list->head = NULL;
    }
    return list;
}

/* Adds a node to the front of the list with the specified data */
void ll_push(LinkedList *list, void *data) {
    Node *newNode = (Node *)malloc(sizeof(Node));
    if (newNode == NULL) {
        return; // Handle malloc failure
    }
    newNode->data = data;
    newNode->next = list->head;
    list->head = newNode;
}

/* Removes and returns the element at the front of the list */
void *ll_pop(LinkedList *list) {
    if (list->head == NULL) {
        return NULL; // No element to pop
    }
    Node *node = list->head;
    void *data = node->data;
    list->head = node->next;
    free(node);
    return data;
}

/* Adds a node to the end of the list with the specified data */
void ll_append(LinkedList *list, void *data) {
    Node *newNode = (Node *)malloc(sizeof(Node));
    if (newNode == NULL) {
        return; // Handle malloc failure
    }
    newNode->data = data;
    newNode->next = NULL;

    if (list->head == NULL) {
        list->head = newNode;
    } else {
        Node *current = list->head;
        while (current->next != NULL) {
            current = current->next;
        }
        current->next = newNode;
    }
}

/* Removes the node with the specified data using a comparison function */
void *ll_remove(LinkedList *list, void *target, int (*compfunc)(void *, void *)) {
    Node *current = list->head;
    Node *previous = NULL;

    while (current != NULL) {
        if (compfunc(current->data, target) == 0) {
            void *data = current->data;
            if (previous == NULL) {
                list->head = current->next;
            } else {
                previous->next = current->next;
            }
            free(current);
            return data;
        }
        previous = current;
        current = current->next;
    }
    return NULL; // No matching node found
}

/* Searches for the node with the specified data */
void *ll_find(LinkedList *list, void *target, int (*compfunc)(void *, void *)) {
    Node *current = list->head;
    while (current != NULL) {
        if (compfunc(current->data, target) == 0) {
            return current->data;
        }
        current = current->next;
    }
    return NULL; // No matching node found
}

/* Returns the number of nodes in the list */
int ll_size(LinkedList *list) {
    int count = 0;
    Node *current = list->head;
    while (current != NULL) {
        count++;
        current = current->next;
    }
    return count;
}

/* Clears the list and frees all associated memory */
void ll_clear(LinkedList *list, void (*freefunc)(void *)) {
    Node *current = list->head;
    Node *next;

    while (current != NULL) {
        next = current->next;
        freefunc(current->data);
        free(current);
        current = next;
    }
    list->head = NULL;
}

/* Applies the provided function to the data in each node */
void ll_map(LinkedList *list, void (*func)(void *)) {
    Node *current = list->head;
    while (current != NULL) {
        func(current->data);
        current = current->next;
    }
}

/* Deletes the node at the specified position */
void ll_delete(LinkedList *list, int position) {
    if (position < 0 || position >= ll_size(list)) {
        return; // Position out of bounds
    }

    Node *current = list->head;
    Node *previous = NULL;

    for (int i = 0; i < position; i++) {
        previous = current;
        current = current->next;
    }

    if (previous == NULL) {
        list->head = current->next; // Head node is being removed
    } else {
        previous->next = current->next;
    }

    free(current);
}

File: Project5_makhan25/LinkedList.js
/**
 * This file provides the implementation for a generic linked list in JavaScript:
 * - Node structure definition which includes a data pointer and a pointer to the next node.
 * - LinkedList structure definition which includes a head pointer to the first node.
 * 
 * It also includes operations such as:
 * - Creation of a new linked list
 * - Push operation to add an element to the front of the list
 * - Pop operation to remove and return the front element from the list
 * - Append operation to add an element to the end of the list
 * - Remove operation to delete a node based on a provided comparison function and target data
 * - Size operation to get the number of elements in the list
 * - Clear operation to empty the list and reset its size
 * - Map operation to apply a function to each element in the list
 *
 * These operations enable the linked list to handle various types of data flexibly and efficiently.
 *
 * Author: Mahdeen Ahmed Khan Sameer
 * Date: April 10, 2024
 */

class Node {
  constructor(data) {
      this.data = data; // holds the data of the node
      this.next = null; // pointer to the next node in the list
  }
}

class LinkedList {
  constructor() {
      this.head = null; // the first node of the list
      this.size = 0;    // number of nodes in the list
  }

  // Adds an element at the front of the list
  push(data) {
      const newNode = new Node(data);
      newNode.next = this.head;  // point the new node to the current head
      this.head = newNode;       // update the head to be the new node
      this.size++;
  }

  // Removes the front element and returns its data
  pop() {
      if (!this.head) return null; // if the list is empty, return null
      const data = this.head.data; // capture the data of the head
      this.head = this.head.next;  // update the head to the next node
      this.size--;
      return data;
  }

  // Adds an element at the end of the list
  append(data) {
      const newNode = new Node(data);
      if (!this.head) {
          this.head = newNode; // if the list is empty, new node becomes the head
      } else {
          let current = this.head;
          while (current.next) {
              current = current.next; // traverse to the end of the list
          }
          current.next = newNode; // link the last node to the new node
      }
      this.size++;
  }

  // Removes the node with data matching the target
  remove(target, compfunc) {
      let current = this.head;
      let prev = null;

      while (current) {
          if (compfunc(current.data, target)) { // comparison function returns true if match
              if (!prev) {
                  this.head = current.next; // remove head if it's a match
              } else {
                  prev.next = current.next; // unlink the current node from the list
              }
              this.size--;
              return current.data; // return the data of the removed node
          }
          prev = current; // move prev and current pointers forward
          current = current.next;
      }
      return null; // if no match found, return null
  }

  // Finds and returns data based on a match with the target
  find(target, compfunc) {
      let current = this.head;

      while (current) {
          if (compfunc(current.data, target)) {
              return current.data; // return the data if a match is found
          }
          current = current.next;
      }
      return null; // if no match found, return null
  }

  // Clears the linked list
  clear() {
      this.head = null; // reset the head pointer
      this.size = 0;    // reset the size of the list
  }

  // Applies a function to each element in the list
  map(mapfunc) {
      let current = this.head;

      while (current) {
          mapfunc(current.data); // apply the function to each node's data
          current = current.next;
      }
  }
}

module.exports = LinkedList; // allows the LinkedList class to be imported in other files
